src/System.Security.Principal.Windows/src/System/Security/Principal/SID.cs

```diff
@@ -792,21 +792,44 @@ public override string ToString()
         {
             if (_sddlForm == null)
             {
-                StringBuilder result = new StringBuilder();
-
                 //
-                // Typecasting of _IdentifierAuthority to a long below is important, since
+                // Typecasting of _IdentifierAuthority to a ulong below is important, since
                 // otherwise you would see this: "S-1-NTAuthority-32-544"
                 //
 
-                result.Append("S-1-").Append((long)_identifierAuthority);
-
+#if netcoreapp20 || netstandard
+                StringBuilder result = new StringBuilder();
+                result.Append("S-1-").Append((ulong)_identifierAuthority);
                 for (int i = 0; i < SubAuthorityCount; i++)
                 {
                     result.Append('-').Append((uint)(_subAuthorities[i]));
                 }
-
                 _sddlForm = result.ToString();
+#else
+                // length of buffer calculation
```

Comment:
Do we have tests in the repo for SecurityIdentifier, and in particular, SecurityIdentifer.ToString()?  As part of changing this code, it'd be great to ensure we have a comprehensive set of tests for ToString.

Reply:
There aren't any tests that cover it. It uses pinvoke on all construction paths so there's no static test other than well known identifiers that could be guaranteed to work. In my test harness I made sure I checked all output against the reference implementation of ToString().

Reply:
A test that confirms that the current user's SID starts with `S-1-5-`, does not end with a -, and everything delimited by a - is CultureInfo.InvariantCulture parsable into a uint seems valid.

Further, it could invoke `%WINDIR%\System32\whoami.exe /user`, split on space, and confirm that "get my SID".ToString() produces the same answer (which ensures RIDs are in the proper order).

Reply:
Are there any tests that do a similar commandline invocation and parse that spring to mind which i can copy code from?

Reply:
https://github.com/dotnet/corefx/blob/776a2da1930cbf08442740be1c13691efb3246f4/src/System.Private.Xml/tests/Xslt/XsltCompiler/XsltCommon.cs#L269-L315 invokes peverify.exe and looks for "PASS" in the output.

That's the closest variant I found with a quick pass.

Reply:
Great, thanks for that. Tests added and pushed.